



<HTML><HEAD>

<TITLE>December 1997/Error Handling with C++ Exceptions, Part 2/Listing 3</TITLE></HEAD>
<BODY BACKGROUND="" BGCOLOR="#FFFFFF" TEXT="#000000">
<h4>Listing 3:	Illustrates the principle of "Resource Allocation is
Initialization"</h4>

<pre>
// destroy5.cpp
#include &lt;stdio.h&gt;

main()
{
    void f(const char*);
    try
    {
        f("file1.dat");
    }
    catch(int)
    {
        puts("Caught exception");
    }
}

void f(const char* fname)
{
    class File
    {
        FILE* f;
    public:
        File(const char* fname,
             const char* mode)
        {
            f = fopen(fname, mode);
        }
        ~File()
        {
            fclose(f);
            puts("File closed");
        }
    };

    File x(fname,"r");
    throw 1;
}

// Output:
File closed
Caught exception
//End of File
</pre>
</body></html>
